src/Dapr.PluggableComponents.AspNetCore/DaprPluggableComponentsApplication.cs

```diff
@@ -0,0 +1,169 @@
+    }
+
+    private void AddRelatedStateStoreServices<TStateStore>()
+    {
+        if (typeof(TStateStore).IsAssignableTo(typeof(IQueryableStateStore)))
+        {
+            this.AddRelatedService<IQueryableStateStore, IStateStore, QueryableStateStoreAdaptor>();
+        }
+
+        if (typeof(TStateStore).IsAssignableTo(typeof(ITransactionalStateStore)))
+        {
+            this.AddRelatedService<ITransactionalStateStore, IStateStore, TransactionalStateStoreAdaptor>();
+        }
+    }
+
+    #endregion
+
+    public void Run()
+    {
+        this.CreateApplication().Run();        
+    }
+
+    public Task RunAsync()
+    {
+        return this.CreateApplication().RunAsync();
+    }
+
+    private void AddComponent<TComponentType, TComponentImpl, TAdaptor>(Func<IServiceProvider, string?, TComponentImpl> pubSubFactory)
+        where TComponentImpl : class, TComponentType
+        where TAdaptor : class
+    {
+        this.ConfigureApplicationBuilder(
+            builder =>
+            {
+                builder.Services.AddSingleton<IDaprPluggableComponentProvider<TComponentType>>(serviceProvider => new MultiplexedComponentProvider<TComponentType>(serviceProvider, pubSubFactory));
+            });
+
+        this.ConfigureApplication(
+            app =>
+            {
+                app.MapDaprPluggableComponent<TAdaptor>();                
+            });
+    }
+
+    private void AddComponent<TComponentType, TComponentImpl, TAdaptor>()
+        where TComponentType : class
+        where TComponentImpl : class, TComponentType
+        where TAdaptor : class
+    {
+        this.ConfigureApplicationBuilder(
+            builder =>
+            {
+                builder.Services.AddSingleton<TComponentType, TComponentImpl>();
+                builder.Services.AddSingleton<IDaprPluggableComponentProvider<TComponentType>, SingletonComponentProvider<TComponentType>>();
+            });
+
+        this.ConfigureApplication(
+            app =>
+            {
+                app.MapDaprPluggableComponent<TAdaptor>();
+            });
+    }
+
+    private void AddRelatedService<TComponent, TComponentImpl, TAdaptor>()
+        where TComponent : class
+        where TAdaptor : class
+    {
+        this.ConfigureApplicationBuilder(
+            builder =>
+            {
+                builder.Services.AddSingleton<IDaprPluggableComponentProvider<TComponent>, DelegatedComponentProvider<TComponent, TComponentImpl>>();
+            });
+
+        this.ConfigureApplication(
+            app =>
+            {
+                app.MapDaprPluggableComponent<TAdaptor>();
+            });
+    }
+
+    private void ConfigureApplication(Action<WebApplication> configurer)
+    {
+        this.appActions.Add(configurer);
+    }
+
+    private void ConfigureApplicationBuilder(Action<WebApplicationBuilder> configurer)
+    {
+        this.builderActions.Add(configurer);
+    }
+
+    private WebApplication CreateApplication()
```

Comment:
I agree that the simplest scenarios shouldn't *require* knowledge of or explicit configuration of that underlying ASP.NET infrastructure, or the use of gRPC at all.

However, given that Dapr Pluggable Components *is* hosted on ASP.NET infrastructure, and then also likely written by developers used to ASP.NET infrastructure, I think it's imperative to provide those developers with all the normal trappings of ASP.NET development (e.g. configuration, service registration, logging, etc.).  This means exposing the `WebApplicationBuilder` and `WebApplication` types directly as many extension methods operate on that level.  Unfortunately, those types aren't amenable to extension (e.g. allowing a `DaprApplicationBuilder` or `DaprApplication` types) that might otherwise eliminate the need to expose those types in other ways.

My goal was to expose a couple "layers" of APIs, a "high level" API that takes care of all Dapr-related registration, which sit upon "middle level" APIs that do specific pieces of Dapr-related registration but can be used to add flexibility when the simple scenarios don't work, and then expose the "low level" ASP.NET APIs to give developers access to both their usual configuration APIs as well as allow flexibility for whatever else might be needed.

Unfortunately, there's no good way to partition the things that we expect developer to need to do (e.g. configuration, logging, and service registration) from other things we don't really expect developers to need to do (e.g. add middleware, expose other hosts and routes, etc.); I don't think it'd be feasible to try to route the all of the former through Dapr SDK methods in order to disallow the latter.

Reply:
I am still going through this PR. That said, with only these 2 comments as context my take is: I think the goals of these SDKs is to be opinionated and err on the side of simplicity from the end-user PoV than having them too configurable.

Objectively, these SDKs are not doing much: they set up a gRPC server to listen on an Unix Domain Socket and define mappings from Protocol Buffer models to more language-friendly models. If a user wants more control they can write and instrument their own gRPC server (which is not much code anyway, specially if they use ours as a base) and even re-use the mappings we provide or write their own.

Reply:
I agree with @tmacam, to make a fair comparison, think about how many of these things that you mentioned a built-in component can do.

I guess almost none of them to be honest, when creating built-in components the user has no ability to deal with service lifecycle, and other framework-specific things. Especially because they are no written on dapr/dapr repository, they reside in the components-contrib repository.

To be fair, they receive a log object as a parameter, and to keep this similarly we should provide a dapr.logger by default as well. 

I totally understand your points but I tend to share the Tiago’s opinion about having opinionated SDKs as much as possible and let the users that don’t need or don’t want that level of abstraction, switch to a raw gRPC implementation.

Reply:
I think it's possible to be *both* highly-opinionated *and* flexible.  ASP.NET itself is a highly-opinionated framework--consider the shift being made from controllers to minimal APIs today with .NET 6/7--but still takes a layered and composable approach, whether you stick to the "old way" or use the "new way" or some combination of both.

While I totally agree that it should be possible to build a pluggable component without seeing any of the underlying ASP.NET or gRPC infrastructure, I'm fairly adamant that developers using this SDK will *expect* to have access to certain core ASP.NET infrastructure, as they are likely to already be ASP.NET developers and used to that style of development.  It would be a shame, in my opinion, to force developers to "start from scratch" just because they wanted to make use of ASP.NET's dependency injection or configuration or logging, or even something slightly more complex such as expose an additional control- or dataplane or dashboard from their component.  Sure, we could funnel DI or logging through our own API layer, but then we take on the burden of matching underlying APIs and maintaining semantics.

My current model allows for a pluggable component as minimal as this, without ever directly exposing ASP.NET or gRPC, but where certain levers still do exist to customize the runtime environment where developers need that flexibility:

```csharp
var app = DaprPluggableComponentsApplication.Create();

app.RegisterService(
    "mystatestore",
    serviceBuilder =>
    {
        serviceBuilder.RegisterStateStore<MyStateStore>();
    });

app.Run();

class MyStateStore : IStateStore
{
    // ...
}
```

Reply:
> I think it's possible to be both highly-opinionated and flexible
I know that, my point is: this flexibility now can cost harder maintenance later

I'm afraid that this is a blocker for the future that we already shape to everything being WASM-compliant. Make this API thin and simple also enables future migrations or allows better controls over pluggable components. 

Also, by reducing from "you can do everything" to "lets do only what is permitted" brings us more control to decide when/what/how the code should be executed.

Even without considering this, I'm a huge fan of enabling things progressively, so that we would avoid breaking changes - which is really annoying even considering sandbox projects. Removing things from the user is really painful, add new things, generally feels like features.

>  It would be a shame, in my opinion, to force developers to "start from scratch" just because they wanted to make use of ASP.NET's dependency injection or configuration or logging, or even something slightly more complex such as expose an additional control- or dataplane or dashboard from their component. Sure, we could funnel DI or logging through our own API layer, but then we take on the burden of matching underlying APIs and maintaining semantics.

Can we discover exactly which are the subset of features that you would consider extremely necessary to build a pluggable components that depends on the framework?

Reply:
> I'm afraid that this is a blocker for the future that we already shape to everything being WASM-compliant. Make this API thin and simple also enables future migrations or allows better controls over pluggable components.

I'd be curious to understand more about the effort to shape things in a WASM-compliant way.

> Can we discover exactly which are the subset of features that you would consider extremely necessary to build a pluggable components that depends on the framework?

At a minimum, I think that would be DI registration and logging, with configuration probably not far behind.

Reply:
Here's what I propose as a compromise:

 - Eliminate (for now) the `DaprPluggableComponentsApplicationOptions` (and its access to `WebApplicationBuilder` and `WebApplication`)
 - Expose the `Services`, `Configuration`, and `Logging` properties of `WebApplicationBuilder` as passthrough properties on `DaprPluggableComponentsApplication` itself

This allows developers access to DI, configuration, and logging APIs as they'd expect in an ASP.NET application, but without the access to the `WebHost` and `WebApplication` to do crazy things like add other endpoints, etc.  I still think there's valid scenarios for that, but we can wait for actual users to tell us that first.  If that happens, we can easily add the options type back.

Each of the types of the `Services`, `Configuration`, and `Logging` properties are from their respective abstract MS libraries--that is, not ASP.NET specific, so ASP.NET types won't be leaked into the SDK (though semantics around DI scopes will still match ASP.NET).